src/modules/content-s3/content-s3.c

```diff
@@ -27,6 +33,217 @@ struct content_s3 {
+        free (ctx->bucket);
+        free (ctx->hostname);
+        free (ctx->access_key);
+        free (ctx->secret_key);
+        free (ctx);
+        errno = saved_errno;
+    }
+}
+
+/* Destroy module context.
+ */
+static void content_s3_destroy (struct content_s3 *ctx)
+{
+    if (ctx) {
+        int saved_errno = errno;
+        flux_msg_handler_delvec (ctx->handlers);
+        s3_config_destroy (ctx->cfg);
+        free (ctx);
+        errno = saved_errno;
+    }
+
+    s3_cleanup ();
+}
+
+static int parse_credentials (struct s3_config *cfg,
+                              const char *cred_file,
+                              char *errbuff,
+                              int eb_size)
+{
+    struct tomltk_error toml_error;
+    toml_table_t *tbl;
+    const char *raw;
+    char *access_key;
+    char *secret_key;
+    int saved_errno;
+
+    if (!(tbl = tomltk_parse_file (cred_file, &toml_error))) {
+        errno = EINVAL;
+        snprintf (errbuff, eb_size, "toml parse failed: %s", toml_error.errbuf);
+        goto error;
+    }
+
+    if (!(raw = toml_raw_in (tbl, "secret-access-key"))) {
+        errno = EINVAL;
+        snprintf (errbuff, eb_size, "failed to parse secret key");
+        goto error;
+    }
+
+    if (toml_rtos (raw, &secret_key)) {
+        errno = EINVAL;
+        snprintf (errbuff, eb_size, "failed to parse secret key");
+        goto error;
+    }
+
+    if (!(raw = toml_raw_in (tbl, "access-key-id"))) {
+        free (secret_key);
+        errno = EINVAL;
+        snprintf (errbuff, eb_size, "failed to parse access key");
+        goto error;
+    }
+
+    if (toml_rtos (raw, &access_key)) {
+        free (secret_key);
+        errno = EINVAL;
+        snprintf (errbuff, eb_size, "failed to parse access key");
+        goto error;
+    }
+
+    cfg->secret_key = secret_key;
+    cfg->access_key = access_key;
+
+    return 0;
+
+error:
+    saved_errno = errno;
+    toml_free (tbl);
+    errno = saved_errno;
+    return -1;
+}
+
+static char *hostport (const char *host, const char *port)
+{
+    char *s;
+    if (port == 0) {
+        if (!(s = strdup (host)))
+            return NULL;
+    }
+    else {
+        if (asprintf (&s, "%s:%d", host, port) < 0)
+            return NULL;
+    }
+    return s;
+}
+
+static struct s3_config *parse_config (const flux_conf_t *conf,
+                                       char *errbuff,
+                                       int eb_size)
+{
+    struct s3_config *cfg;
+    flux_conf_error_t error;
+    const char *uri = NULL;
+    const char *bucket = NULL;
+    const char *cred_file = NULL;
+    int is_virtual_host = 0;
+    char tmp[6];
+    struct yuarel yuri;
+    char *cpy = NULL;
+    int saved_errno;
+
+    if (!(cfg = calloc (1, sizeof (*cfg))))
+        return NULL;
+
+    cfg->retries = 5;
+    cfg->is_secure = 0;
+
+    if (flux_conf_unpack (conf,
+                          &error,
+                          "{s:{s:s, s:s, s:s, s?:b} !}",
```

Comment:
Was playing around with some tests and I happened to notice that the `!` ought to be inside inner braces.  It's supposed to catch keys in the `[content-s3]` table that are unknown.